graph/pull_v2.go

```diff
@@ -223,6 +223,9 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (verified bool, err error)
 	go func() {
 		if _, err := io.Copy(out, pipeReader); err != nil {
 			logrus.Errorf("error copying from layer download progress reader: %s", err)
+			if _, err := io.Copy(ioutil.Discard, pipeReader); err != nil {
```

Comment:
This was implemented with https://github.com/docker/docker/pull/15353. The original idea was to cancel the download by controlling the download pipe, not the progress reporting pipe. Looks like I missed this detail in review. This lets you cancel the `io.Copy` in https://github.com/docker/docker/pull/15353/files#diff-7e2a216082b6bd557c7ecebd68305e2eL174 and stop the download (ack! that error should be checked, as well).

Studying this again, the use of `pipeReader` is correct. Rather than sending to discard, it should be sufficient to call `CloseWithError` on the `pipeReader`.